From 3e0b604e77857059b18b93d83d6774e52ef3acf4 Mon Sep 17 00:00:00 2001
From: Jason Dere <jdere@apache.org>
Date: Tue, 7 Apr 2015 22:17:50 +0000
Subject: [PATCH 0423/1164] CDH-33773: HIVE-10178: DateWritable incorrectly
 calculates daysSinceEpoch for negative Unix time
 (Alex Pivovarov via Jason Dere)

Change-Id: If6dd225808c96946b11f0a11b77c60cc8495a833
git-svn-id: https://svn.apache.org/repos/asf/hive/trunk@1671961 13f79535-47bb-0310-9956-ffa450edef68
---
 .../hive/ql/udf/generic/TestGenericUDFLastDay.java |   41 +-
 .../clientpositive/interval_arithmetic.q.out       |  620 ++++++++++++++++++++
 .../clientpositive/spark/vector_between_in.q.out   |  145 ++---
 .../clientpositive/tez/update_all_types.q.out      |   26 +-
 .../clientpositive/tez/vector_between_in.q.out     |  145 ++---
 .../results/clientpositive/update_all_types.q.out  |   26 +-
 .../results/clientpositive/vector_between_in.q.out |  145 ++---
 .../apache/hadoop/hive/serde2/io/DateWritable.java |   19 +-
 8 files changed, 858 insertions(+), 309 deletions(-)
 create mode 100644 ql/src/test/results/clientpositive/interval_arithmetic.q.out

diff --git a/ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFLastDay.java b/ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFLastDay.java
index a842876..65f706a 100644
--- a/ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFLastDay.java
+++ b/ql/src/test/org/apache/hadoop/hive/ql/udf/generic/TestGenericUDFLastDay.java
@@ -17,16 +17,18 @@
  */
 package org.apache.hadoop.hive.ql.udf.generic;
 
+import java.sql.Timestamp;
+
+import junit.framework.TestCase;
+
 import org.apache.hadoop.hive.ql.metadata.HiveException;
-import org.apache.hadoop.hive.ql.udf.generic.GenericUDF;
 import org.apache.hadoop.hive.ql.udf.generic.GenericUDF.DeferredJavaObject;
 import org.apache.hadoop.hive.ql.udf.generic.GenericUDF.DeferredObject;
+import org.apache.hadoop.hive.serde2.io.TimestampWritable;
 import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
 import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
 import org.apache.hadoop.io.Text;
 
-import junit.framework.TestCase;
-
 public class TestGenericUDFLastDay extends TestCase {
 
   public void testLastDay() throws HiveException {
@@ -65,6 +67,31 @@ public void testLastDay() throws HiveException {
     runAndVerify("2014-01-32 10:30:45", "2014-02-28", udf);
     runAndVerify("01/14/2014 10:30:45", null, udf);
     runAndVerify("2016-02-28T10:30:45", "2016-02-29", udf);
+    // negative Unix time
+    runAndVerifyTs("1966-01-31 00:00:01", "1966-01-31", udf);
+    runAndVerifyTs("1966-01-31 10:00:01", "1966-01-31", udf);
+    runAndVerifyTs("1966-01-31 23:59:59", "1966-01-31", udf);
+  }
+
+  public void testLastDayTs() throws HiveException {
+    GenericUDFLastDay udf = new GenericUDFLastDay();
+    ObjectInspector valueOI0 = PrimitiveObjectInspectorFactory.writableTimestampObjectInspector;
+    ObjectInspector[] arguments = { valueOI0 };
+
+    udf.initialize(arguments);
+    // positive Unix time
+    runAndVerifyTs("2014-01-01 10:30:45", "2014-01-31", udf);
+    runAndVerifyTs("2014-01-14 10:30:45", "2014-01-31", udf);
+    runAndVerifyTs("2014-01-31 10:30:45.1", "2014-01-31", udf);
+    runAndVerifyTs("2014-02-02 10:30:45.100", "2014-02-28", udf);
+    runAndVerifyTs("2014-02-28 10:30:45.001", "2014-02-28", udf);
+    runAndVerifyTs("2016-02-03 10:30:45.000000001", "2016-02-29", udf);
+    runAndVerifyTs("2016-02-28 10:30:45", "2016-02-29", udf);
+    runAndVerifyTs("2016-02-29 10:30:45", "2016-02-29", udf);
+    // negative Unix time
+    runAndVerifyTs("1966-01-31 00:00:01", "1966-01-31", udf);
+    runAndVerifyTs("1966-01-31 10:00:01", "1966-01-31", udf);
+    runAndVerifyTs("1966-01-31 23:59:59", "1966-01-31", udf);
   }
 
   private void runAndVerify(String str, String expResult, GenericUDF udf)
@@ -74,4 +101,12 @@ private void runAndVerify(String str, String expResult, GenericUDF udf)
     Text output = (Text) udf.evaluate(args);
     assertEquals("last_day() test ", expResult, output != null ? output.toString() : null);
   }
+
+  private void runAndVerifyTs(String str, String expResult, GenericUDF udf) throws HiveException {
+    DeferredObject valueObj0 = new DeferredJavaObject(str != null ? new TimestampWritable(
+        Timestamp.valueOf(str)) : null);
+    DeferredObject[] args = { valueObj0 };
+    Text output = (Text) udf.evaluate(args);
+    assertEquals("last_day() test ", expResult, output != null ? output.toString() : null);
+  }
 }
diff --git a/ql/src/test/results/clientpositive/interval_arithmetic.q.out b/ql/src/test/results/clientpositive/interval_arithmetic.q.out
new file mode 100644
index 0000000..b6c3fdf
--- /dev/null
+++ b/ql/src/test/results/clientpositive/interval_arithmetic.q.out
@@ -0,0 +1,620 @@
+PREHOOK: query: create table interval_arithmetic_1 (dateval date, tsval timestamp)
+PREHOOK: type: CREATETABLE
+PREHOOK: Output: database:default
+PREHOOK: Output: default@interval_arithmetic_1
+POSTHOOK: query: create table interval_arithmetic_1 (dateval date, tsval timestamp)
+POSTHOOK: type: CREATETABLE
+POSTHOOK: Output: database:default
+POSTHOOK: Output: default@interval_arithmetic_1
+PREHOOK: query: insert overwrite table interval_arithmetic_1
+  select cast(ctimestamp1 as date), ctimestamp1 from alltypesorc
+PREHOOK: type: QUERY
+PREHOOK: Input: default@alltypesorc
+PREHOOK: Output: default@interval_arithmetic_1
+POSTHOOK: query: insert overwrite table interval_arithmetic_1
+  select cast(ctimestamp1 as date), ctimestamp1 from alltypesorc
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@alltypesorc
+POSTHOOK: Output: default@interval_arithmetic_1
+POSTHOOK: Lineage: interval_arithmetic_1.dateval EXPRESSION [(alltypesorc)alltypesorc.FieldSchema(name:ctimestamp1, type:timestamp, comment:null), ]
+POSTHOOK: Lineage: interval_arithmetic_1.tsval SIMPLE [(alltypesorc)alltypesorc.FieldSchema(name:ctimestamp1, type:timestamp, comment:null), ]
+PREHOOK: query: -- interval year-month arithmetic
+explain
+select
+  dateval,
+  dateval - interval '2-2' year to month,
+  dateval - interval '-2-2' year to month,
+  dateval + interval '2-2' year to month,
+  dateval + interval '-2-2' year to month,
+  - interval '2-2' year to month + dateval,
+  interval '2-2' year to month + dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: -- interval year-month arithmetic
+explain
+select
+  dateval,
+  dateval - interval '2-2' year to month,
+  dateval - interval '-2-2' year to month,
+  dateval + interval '2-2' year to month,
+  dateval + interval '-2-2' year to month,
+  - interval '2-2' year to month + dateval,
+  interval '2-2' year to month + dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: dateval (type: date), (dateval - 2-2) (type: date), (dateval - -2-2) (type: date), (dateval + 2-2) (type: date), (dateval + -2-2) (type: date), (-2-2 + dateval) (type: date), (2-2 + dateval) (type: date)
+              outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  dateval,
+  dateval - interval '2-2' year to month,
+  dateval - interval '-2-2' year to month,
+  dateval + interval '2-2' year to month,
+  dateval + interval '-2-2' year to month,
+  - interval '2-2' year to month + dateval,
+  interval '2-2' year to month + dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  dateval,
+  dateval - interval '2-2' year to month,
+  dateval - interval '-2-2' year to month,
+  dateval + interval '2-2' year to month,
+  dateval + interval '-2-2' year to month,
+  - interval '2-2' year to month + dateval,
+  interval '2-2' year to month + dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31	1967-10-31	1972-02-29	1972-02-29	1967-10-31	1967-10-31	1972-02-29
+NULL	NULL	NULL	NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  dateval,
+  dateval - date '1999-06-07',
+  date '1999-06-07' - dateval,
+  dateval - dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  dateval,
+  dateval - date '1999-06-07',
+  date '1999-06-07' - dateval,
+  dateval - dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: dateval (type: date), (dateval - 1999-06-07) (type: interval_day_time), (1999-06-07 - dateval) (type: interval_day_time), (dateval - dateval) (type: interval_day_time)
+              outputColumnNames: _col0, _col1, _col2, _col3
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  dateval,
+  dateval - date '1999-06-07',
+  date '1999-06-07' - dateval,
+  dateval - dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  dateval,
+  dateval - date '1999-06-07',
+  date '1999-06-07' - dateval,
+  dateval - dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31	-10749 23:00:00.000000000	10749 23:00:00.000000000	0 00:00:00.000000000
+NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  tsval,
+  tsval - interval '2-2' year to month,
+  tsval - interval '-2-2' year to month,
+  tsval + interval '2-2' year to month,
+  tsval + interval '-2-2' year to month,
+  - interval '2-2' year to month + tsval,
+  interval '2-2' year to month + tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  tsval,
+  tsval - interval '2-2' year to month,
+  tsval - interval '-2-2' year to month,
+  tsval + interval '2-2' year to month,
+  tsval + interval '-2-2' year to month,
+  - interval '2-2' year to month + tsval,
+  interval '2-2' year to month + tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: tsval (type: timestamp), (tsval - 2-2) (type: timestamp), (tsval - -2-2) (type: timestamp), (tsval + 2-2) (type: timestamp), (tsval + -2-2) (type: timestamp), (-2-2 + tsval) (type: timestamp), (2-2 + tsval) (type: timestamp)
+              outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  tsval,
+  tsval - interval '2-2' year to month,
+  tsval - interval '-2-2' year to month,
+  tsval + interval '2-2' year to month,
+  tsval + interval '-2-2' year to month,
+  - interval '2-2' year to month + tsval,
+  interval '2-2' year to month + tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  tsval,
+  tsval - interval '2-2' year to month,
+  tsval - interval '-2-2' year to month,
+  tsval + interval '2-2' year to month,
+  tsval + interval '-2-2' year to month,
+  - interval '2-2' year to month + tsval,
+  interval '2-2' year to month + tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31 15:59:46.674	1967-10-31 15:59:46.674	1972-02-29 15:59:46.674	1972-02-29 15:59:46.674	1967-10-31 15:59:46.674	1967-10-31 15:59:46.674	1972-02-29 15:59:46.674
+NULL	NULL	NULL	NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  interval '2-2' year to month + interval '3-3' year to month,
+  interval '2-2' year to month - interval '3-3' year to month
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  interval '2-2' year to month + interval '3-3' year to month,
+  interval '2-2' year to month - interval '3-3' year to month
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-0 is a root stage
+
+STAGE PLANS:
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        TableScan
+          alias: interval_arithmetic_1
+          Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: COMPLETE
+          Select Operator
+            expressions: 5-5 (type: interval_year_month), -1-1 (type: interval_year_month)
+            outputColumnNames: _col0, _col1
+            Statistics: Num rows: 12288 Data size: 0 Basic stats: PARTIAL Column stats: COMPLETE
+            Limit
+              Number of rows: 2
+              Statistics: Num rows: 2 Data size: 0 Basic stats: PARTIAL Column stats: COMPLETE
+              ListSink
+
+PREHOOK: query: select
+  interval '2-2' year to month + interval '3-3' year to month,
+  interval '2-2' year to month - interval '3-3' year to month
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  interval '2-2' year to month + interval '3-3' year to month,
+  interval '2-2' year to month - interval '3-3' year to month
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+5-5	-1-1
+5-5	-1-1
+PREHOOK: query: -- interval day-time arithmetic
+explain
+select
+  dateval,
+  dateval - interval '99 11:22:33.123456789' day to second,
+  dateval - interval '-99 11:22:33.123456789' day to second,
+  dateval + interval '99 11:22:33.123456789' day to second,
+  dateval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + dateval,
+  interval '99 11:22:33.123456789' day to second + dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: -- interval day-time arithmetic
+explain
+select
+  dateval,
+  dateval - interval '99 11:22:33.123456789' day to second,
+  dateval - interval '-99 11:22:33.123456789' day to second,
+  dateval + interval '99 11:22:33.123456789' day to second,
+  dateval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + dateval,
+  interval '99 11:22:33.123456789' day to second + dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: dateval (type: date), (dateval - 99 11:22:33.123456789) (type: timestamp), (dateval - -99 11:22:33.123456789) (type: timestamp), (dateval + 99 11:22:33.123456789) (type: timestamp), (dateval + -99 11:22:33.123456789) (type: timestamp), (-99 11:22:33.123456789 + dateval) (type: timestamp), (99 11:22:33.123456789 + dateval) (type: timestamp)
+              outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  dateval,
+  dateval - interval '99 11:22:33.123456789' day to second,
+  dateval - interval '-99 11:22:33.123456789' day to second,
+  dateval + interval '99 11:22:33.123456789' day to second,
+  dateval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + dateval,
+  interval '99 11:22:33.123456789' day to second + dateval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  dateval,
+  dateval - interval '99 11:22:33.123456789' day to second,
+  dateval - interval '-99 11:22:33.123456789' day to second,
+  dateval + interval '99 11:22:33.123456789' day to second,
+  dateval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + dateval,
+  interval '99 11:22:33.123456789' day to second + dateval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31	1969-09-22 13:37:26.876543211	1970-04-09 11:22:33.123456789	1970-04-09 11:22:33.123456789	1969-09-22 13:37:26.876543211	1969-09-22 13:37:26.876543211	1970-04-09 11:22:33.123456789
+NULL	NULL	NULL	NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  dateval,
+  tsval,
+  dateval - tsval,
+  tsval - dateval,
+  tsval - tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  dateval,
+  tsval,
+  dateval - tsval,
+  tsval - dateval,
+  tsval - tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: dateval (type: date), tsval (type: timestamp), (dateval - tsval) (type: interval_day_time), (tsval - dateval) (type: interval_day_time), (tsval - tsval) (type: interval_day_time)
+              outputColumnNames: _col0, _col1, _col2, _col3, _col4
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  dateval,
+  tsval,
+  dateval - tsval,
+  tsval - dateval,
+  tsval - tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  dateval,
+  tsval,
+  dateval - tsval,
+  tsval - dateval,
+  tsval - tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31	1969-12-31 15:59:46.674	-0 15:59:47.674000000	0 15:59:47.674000000	0 00:00:00.000000000
+NULL	NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  tsval,
+  tsval - interval '99 11:22:33.123456789' day to second,
+  tsval - interval '-99 11:22:33.123456789' day to second,
+  tsval + interval '99 11:22:33.123456789' day to second,
+  tsval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + tsval,
+  interval '99 11:22:33.123456789' day to second + tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  tsval,
+  tsval - interval '99 11:22:33.123456789' day to second,
+  tsval - interval '-99 11:22:33.123456789' day to second,
+  tsval + interval '99 11:22:33.123456789' day to second,
+  tsval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + tsval,
+  interval '99 11:22:33.123456789' day to second + tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-1 is a root stage
+  Stage-0 depends on stages: Stage-1
+
+STAGE PLANS:
+  Stage: Stage-1
+    Map Reduce
+      Map Operator Tree:
+          TableScan
+            alias: interval_arithmetic_1
+            Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+            Select Operator
+              expressions: tsval (type: timestamp), (tsval - 99 11:22:33.123456789) (type: timestamp), (tsval - -99 11:22:33.123456789) (type: timestamp), (tsval + 99 11:22:33.123456789) (type: timestamp), (tsval + -99 11:22:33.123456789) (type: timestamp), (-99 11:22:33.123456789 + tsval) (type: timestamp), (99 11:22:33.123456789 + tsval) (type: timestamp)
+              outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5, _col6
+              Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: NONE
+              Limit
+                Number of rows: 2
+                Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                File Output Operator
+                  compressed: false
+                  Statistics: Num rows: 2 Data size: 52 Basic stats: COMPLETE Column stats: NONE
+                  table:
+                      input format: org.apache.hadoop.mapred.TextInputFormat
+                      output format: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
+                      serde: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
+
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        ListSink
+
+PREHOOK: query: select
+  tsval,
+  tsval - interval '99 11:22:33.123456789' day to second,
+  tsval - interval '-99 11:22:33.123456789' day to second,
+  tsval + interval '99 11:22:33.123456789' day to second,
+  tsval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + tsval,
+  interval '99 11:22:33.123456789' day to second + tsval
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  tsval,
+  tsval - interval '99 11:22:33.123456789' day to second,
+  tsval - interval '-99 11:22:33.123456789' day to second,
+  tsval + interval '99 11:22:33.123456789' day to second,
+  tsval + interval '-99 11:22:33.123456789' day to second,
+  -interval '99 11:22:33.123456789' day to second + tsval,
+  interval '99 11:22:33.123456789' day to second + tsval
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+1969-12-31 15:59:46.674	1969-09-23 05:37:13.550543211	1970-04-10 03:22:19.797456789	1970-04-10 03:22:19.797456789	1969-09-23 05:37:13.550543211	1969-09-23 05:37:13.550543211	1970-04-10 03:22:19.797456789
+NULL	NULL	NULL	NULL	NULL	NULL	NULL
+PREHOOK: query: explain
+select
+  interval '99 11:22:33.123456789' day to second + interval '10 9:8:7.123456789' day to second,
+  interval '99 11:22:33.123456789' day to second - interval '10 9:8:7.123456789' day to second
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+POSTHOOK: query: explain
+select
+  interval '99 11:22:33.123456789' day to second + interval '10 9:8:7.123456789' day to second,
+  interval '99 11:22:33.123456789' day to second - interval '10 9:8:7.123456789' day to second
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+STAGE DEPENDENCIES:
+  Stage-0 is a root stage
+
+STAGE PLANS:
+  Stage: Stage-0
+    Fetch Operator
+      limit: 2
+      Processor Tree:
+        TableScan
+          alias: interval_arithmetic_1
+          Statistics: Num rows: 12288 Data size: 326837 Basic stats: COMPLETE Column stats: COMPLETE
+          Select Operator
+            expressions: 109 20:30:40.246913578 (type: interval_day_time), 89 02:14:26.000000000 (type: interval_day_time)
+            outputColumnNames: _col0, _col1
+            Statistics: Num rows: 12288 Data size: 0 Basic stats: PARTIAL Column stats: COMPLETE
+            Limit
+              Number of rows: 2
+              Statistics: Num rows: 2 Data size: 0 Basic stats: PARTIAL Column stats: COMPLETE
+              ListSink
+
+PREHOOK: query: select
+  interval '99 11:22:33.123456789' day to second + interval '10 9:8:7.123456789' day to second,
+  interval '99 11:22:33.123456789' day to second - interval '10 9:8:7.123456789' day to second
+from interval_arithmetic_1
+limit 2
+PREHOOK: type: QUERY
+PREHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+POSTHOOK: query: select
+  interval '99 11:22:33.123456789' day to second + interval '10 9:8:7.123456789' day to second,
+  interval '99 11:22:33.123456789' day to second - interval '10 9:8:7.123456789' day to second
+from interval_arithmetic_1
+limit 2
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@interval_arithmetic_1
+#### A masked pattern was here ####
+109 20:30:40.246913578	89 02:14:26.000000000
+109 20:30:40.246913578	89 02:14:26.000000000
+PREHOOK: query: drop table interval_arithmetic_1
+PREHOOK: type: DROPTABLE
+PREHOOK: Input: default@interval_arithmetic_1
+PREHOOK: Output: default@interval_arithmetic_1
+POSTHOOK: query: drop table interval_arithmetic_1
+POSTHOOK: type: DROPTABLE
+POSTHOOK: Input: default@interval_arithmetic_1
+POSTHOOK: Output: default@interval_arithmetic_1
diff --git a/ql/src/test/results/clientpositive/spark/vector_between_in.q.out b/ql/src/test/results/clientpositive/spark/vector_between_in.q.out
index 3cdb0ca..75520f4 100644
--- a/ql/src/test/results/clientpositive/spark/vector_between_in.q.out
+++ b/ql/src/test/results/clientpositive/spark/vector_between_in.q.out
@@ -455,36 +455,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-07-14
 1969-07-14
 1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
 1969-10-26
 1969-10-26
 1969-10-26
@@ -507,7 +477,7 @@ POSTHOOK: query: SELECT COUNT(*) FROM decimal_date_test WHERE cdate NOT IN (CAST
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-5996
+6026
 PREHOOK: query: SELECT cdecimal1 FROM decimal_date_test WHERE cdecimal1 IN (2365.8945945946, 881.0135135135, -3367.6517567568) ORDER BY cdecimal1
 PREHOOK: type: QUERY
 PREHOOK: Input: default@decimal_date_test
@@ -550,6 +520,45 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-30
 1969-12-30
 1969-12-30
+1969-12-30
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
 1969-12-31
 1969-12-31
 1969-12-31
@@ -559,53 +568,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-31
 1969-12-31
 1969-12-31
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
 1970-01-01
 1970-01-01
 1970-01-01
@@ -642,22 +604,23 @@ POSTHOOK: query: SELECT cdate FROM decimal_date_test WHERE cdate NOT BETWEEN CAS
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-1968-04-07
+1968-04-06
+1968-04-08
 1968-04-09
-1968-04-10
-1968-04-14
-1968-04-16
-1968-04-16
-1968-04-19
-1968-04-23
+1968-04-13
+1968-04-15
+1968-04-15
+1968-04-18
+1968-04-22
+1968-04-24
 1968-04-25
 1968-04-26
-1968-04-27
-1968-04-27
-1968-04-27
-1968-04-29
-1968-04-29
-1968-04-29
+1968-04-26
+1968-04-26
+1968-04-28
+1968-04-28
+1968-04-28
+1968-04-28
 1968-04-29
 1968-04-30
 1971-09-02
diff --git a/ql/src/test/results/clientpositive/tez/update_all_types.q.out b/ql/src/test/results/clientpositive/tez/update_all_types.q.out
index eba4dde..0e268b9 100644
--- a/ql/src/test/results/clientpositive/tez/update_all_types.q.out
+++ b/ql/src/test/results/clientpositive/tez/update_all_types.q.out
@@ -88,14 +88,14 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -51	NULL	-1071480828	-1401575336	-51.0	NULL	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-741	-1070883071	-1645852809	NULL	-741.0	NULL	NULL	1970-01-01	0ruyd6Y50JpdGRf6HqD	0ruyd6Y50JpdGRf6HqD	xH7445Rals48VOulSyR5F               	false
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-741	-1070883071	-1645852809	NULL	-741.0	NULL	NULL	1969-12-31	0ruyd6Y50JpdGRf6HqD	0ruyd6Y50JpdGRf6HqD	xH7445Rals48VOulSyR5F               	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 PREHOOK: query: update acid_uat set
     ti = 1,
@@ -142,13 +142,13 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -51	NULL	-1071480828	-1401575336	-51.0	NULL	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 1	2	3	4	3.14	6.28	5.99	NULL	2014-09-01	its a beautiful day in the neighbhorhood	a beautiful day for a neighbor	wont you be mine                    	true
 PREHOOK: query: update acid_uat set
@@ -176,12 +176,12 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -102	-51	-1071480828	-1401575336	-51.0	-51.0	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 1	2	3	4	3.14	6.28	5.99	NULL	2014-09-01	its a beautiful day in the neighbhorhood	a beautiful day for a neighbor	wont you be mine                    	true
diff --git a/ql/src/test/results/clientpositive/tez/vector_between_in.q.out b/ql/src/test/results/clientpositive/tez/vector_between_in.q.out
index 651a458..61e215e 100644
--- a/ql/src/test/results/clientpositive/tez/vector_between_in.q.out
+++ b/ql/src/test/results/clientpositive/tez/vector_between_in.q.out
@@ -455,36 +455,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-07-14
 1969-07-14
 1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
 1969-10-26
 1969-10-26
 1969-10-26
@@ -507,7 +477,7 @@ POSTHOOK: query: SELECT COUNT(*) FROM decimal_date_test WHERE cdate NOT IN (CAST
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-5996
+6026
 PREHOOK: query: SELECT cdecimal1 FROM decimal_date_test WHERE cdecimal1 IN (2365.8945945946, 881.0135135135, -3367.6517567568) ORDER BY cdecimal1
 PREHOOK: type: QUERY
 PREHOOK: Input: default@decimal_date_test
@@ -550,6 +520,45 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-30
 1969-12-30
 1969-12-30
+1969-12-30
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
 1969-12-31
 1969-12-31
 1969-12-31
@@ -559,53 +568,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-31
 1969-12-31
 1969-12-31
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
 1970-01-01
 1970-01-01
 1970-01-01
@@ -642,22 +604,23 @@ POSTHOOK: query: SELECT cdate FROM decimal_date_test WHERE cdate NOT BETWEEN CAS
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-1968-04-07
+1968-04-06
+1968-04-08
 1968-04-09
-1968-04-10
-1968-04-14
-1968-04-16
-1968-04-16
-1968-04-19
-1968-04-23
+1968-04-13
+1968-04-15
+1968-04-15
+1968-04-18
+1968-04-22
+1968-04-24
 1968-04-25
 1968-04-26
-1968-04-27
-1968-04-27
-1968-04-27
-1968-04-29
-1968-04-29
-1968-04-29
+1968-04-26
+1968-04-26
+1968-04-28
+1968-04-28
+1968-04-28
+1968-04-28
 1968-04-29
 1968-04-30
 1971-09-02
diff --git a/ql/src/test/results/clientpositive/update_all_types.q.out b/ql/src/test/results/clientpositive/update_all_types.q.out
index eba4dde..0e268b9 100644
--- a/ql/src/test/results/clientpositive/update_all_types.q.out
+++ b/ql/src/test/results/clientpositive/update_all_types.q.out
@@ -88,14 +88,14 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -51	NULL	-1071480828	-1401575336	-51.0	NULL	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-741	-1070883071	-1645852809	NULL	-741.0	NULL	NULL	1970-01-01	0ruyd6Y50JpdGRf6HqD	0ruyd6Y50JpdGRf6HqD	xH7445Rals48VOulSyR5F               	false
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-741	-1070883071	-1645852809	NULL	-741.0	NULL	NULL	1969-12-31	0ruyd6Y50JpdGRf6HqD	0ruyd6Y50JpdGRf6HqD	xH7445Rals48VOulSyR5F               	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 PREHOOK: query: update acid_uat set
     ti = 1,
@@ -142,13 +142,13 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -51	NULL	-1071480828	-1401575336	-51.0	NULL	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 1	2	3	4	3.14	6.28	5.99	NULL	2014-09-01	its a beautiful day in the neighbhorhood	a beautiful day for a neighbor	wont you be mine                    	true
 PREHOOK: query: update acid_uat set
@@ -176,12 +176,12 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@acid_uat
 #### A masked pattern was here ####
 11	NULL	-1073279343	-1595604468	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	oj1YrV5Wa	oj1YrV5Wa	P76636jJ6qM17d7DIy                  	true
-NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1970-01-01	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
+NULL	-7382	-1073051226	-1887561756	NULL	-7382.0	NULL	NULL	1969-12-31	A34p7oRr2WvUJNf	A34p7oRr2WvUJNf	4hA4KQj2vD3fI6gX82220d              	false
 11	NULL	-1072910839	2048385991	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	0iqrc5	0iqrc5	KbaDXiN85adbHRx58v                  	false
-NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1970-01-01	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
-NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1970-01-01	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
+NULL	8373	-1072081801	1864027286	NULL	8373.0	NULL	NULL	1969-12-31	dPkN74F7	dPkN74F7	4KWs6gw7lv2WYd66P                   	true
+NULL	-5470	-1072076362	1864027286	NULL	-5470.0	NULL	NULL	1969-12-31	2uLyD28144vklju213J1mr	2uLyD28144vklju213J1mr	4KWs6gw7lv2WYd66P                   	true
 -102	-51	-1071480828	-1401575336	-51.0	-51.0	-51	1969-12-31 16:00:08.451	NULL	aw724t8c5558x2xneC624	aw724t8c5558x2xneC624	4uE7l74tESBiKfu7c8wM7GA             	true
 8	NULL	-1071363017	1349676361	8.0	NULL	8	1969-12-31 16:00:15.892	NULL	Anj0oF	Anj0oF	IwE1G7Qb0B1NEfV030g                 	true
-NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1970-01-01	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
+NULL	-947	-1070551679	1864027286	NULL	-947.0	NULL	NULL	1969-12-31	iUR3Q	iUR3Q	4KWs6gw7lv2WYd66P                   	false
 11	NULL	-1069736047	-453772520	11.0	NULL	11	1969-12-31 16:00:02.351	NULL	k17Am8uPHWk02cEf1jet	k17Am8uPHWk02cEf1jet	qrXLLNX1                            	true
 1	2	3	4	3.14	6.28	5.99	NULL	2014-09-01	its a beautiful day in the neighbhorhood	a beautiful day for a neighbor	wont you be mine                    	true
diff --git a/ql/src/test/results/clientpositive/vector_between_in.q.out b/ql/src/test/results/clientpositive/vector_between_in.q.out
index c5f3cd1..a9b9a4b 100644
--- a/ql/src/test/results/clientpositive/vector_between_in.q.out
+++ b/ql/src/test/results/clientpositive/vector_between_in.q.out
@@ -399,36 +399,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-07-14
 1969-07-14
 1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-07-14
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
-1969-10-26
 1969-10-26
 1969-10-26
 1969-10-26
@@ -451,7 +421,7 @@ POSTHOOK: query: SELECT COUNT(*) FROM decimal_date_test WHERE cdate NOT IN (CAST
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-5996
+6026
 PREHOOK: query: SELECT cdecimal1 FROM decimal_date_test WHERE cdecimal1 IN (2365.8945945946, 881.0135135135, -3367.6517567568) ORDER BY cdecimal1
 PREHOOK: type: QUERY
 PREHOOK: Input: default@decimal_date_test
@@ -494,6 +464,45 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-30
 1969-12-30
 1969-12-30
+1969-12-30
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
+1969-12-31
 1969-12-31
 1969-12-31
 1969-12-31
@@ -503,53 +512,6 @@ POSTHOOK: Input: default@decimal_date_test
 1969-12-31
 1969-12-31
 1969-12-31
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
-1970-01-01
 1970-01-01
 1970-01-01
 1970-01-01
@@ -586,22 +548,23 @@ POSTHOOK: query: SELECT cdate FROM decimal_date_test WHERE cdate NOT BETWEEN CAS
 POSTHOOK: type: QUERY
 POSTHOOK: Input: default@decimal_date_test
 #### A masked pattern was here ####
-1968-04-07
+1968-04-06
+1968-04-08
 1968-04-09
-1968-04-10
-1968-04-14
-1968-04-16
-1968-04-16
-1968-04-19
-1968-04-23
+1968-04-13
+1968-04-15
+1968-04-15
+1968-04-18
+1968-04-22
+1968-04-24
 1968-04-25
 1968-04-26
-1968-04-27
-1968-04-27
-1968-04-27
-1968-04-29
-1968-04-29
-1968-04-29
+1968-04-26
+1968-04-26
+1968-04-28
+1968-04-28
+1968-04-28
+1968-04-28
 1968-04-29
 1968-04-30
 1971-09-02
diff --git a/serde/src/java/org/apache/hadoop/hive/serde2/io/DateWritable.java b/serde/src/java/org/apache/hadoop/hive/serde2/io/DateWritable.java
index 2b0b550..6cedf4c 100644
--- a/serde/src/java/org/apache/hadoop/hive/serde2/io/DateWritable.java
+++ b/serde/src/java/org/apache/hadoop/hive/serde2/io/DateWritable.java
@@ -25,9 +25,6 @@
 import java.util.TimeZone;
 import java.util.concurrent.TimeUnit;
 
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
-import org.apache.hadoop.hive.serde2.ByteStream.Output;
 import org.apache.hadoop.hive.serde2.ByteStream.RandomAccessOutput;
 import org.apache.hadoop.hive.serde2.lazybinary.LazyBinaryUtils;
 import org.apache.hadoop.hive.serde2.lazybinary.LazyBinaryUtils.VInt;
@@ -44,8 +41,6 @@
  *
  */
 public class DateWritable implements WritableComparable<DateWritable> {
-  private static final Log LOG = LogFactory.getLog(DateWritable.class);
-
   private static final long MILLIS_PER_DAY = TimeUnit.DAYS.toMillis(1);
 
   // Local time zone.
@@ -134,11 +129,21 @@ public static long daysToMillis(int d) {
     return millisUtc - LOCAL_TIMEZONE.get().getOffset(tmp);
   }
 
+  public static int millisToDays(long millisLocal) {
+    long millisUtc = millisLocal + LOCAL_TIMEZONE.get().getOffset(millisLocal);
+    int days;
+    if (millisUtc >= 0L) {
+      days = (int) (millisUtc / MILLIS_PER_DAY);
+    } else {
+      days = (int) ((millisUtc - 86399999) / MILLIS_PER_DAY);
+    }
+    return days;
+  }
+
   public static int dateToDays(Date d) {
     // convert to equivalent time in UTC, then get day offset
     long millisLocal = d.getTime();
-    long millisUtc = millisLocal + LOCAL_TIMEZONE.get().getOffset(millisLocal);
-    return (int)(millisUtc / MILLIS_PER_DAY);
+    return millisToDays(millisLocal);
   }
 
   public void setFromBytes(byte[] bytes, int offset, int length, VInt vInt) {
-- 
1.7.9.5

